### PR TITLE
Add hyperparameter tuning and optimizers for DCGAN

### PR DESCRIPTION
**Pull request by pratik-633**
Implemented hyperparameter tuning for DCGAN training and added optimizers for discriminator and generator.

### CODE REVIEW

File: code/train.py [hunk 1]
```diff
@@ -32,8 +32,24 @@
...
+    
+    params = {
+        'lr': 2e-4,
+        'adam_b1': 0.5,
+        'adam_b2': 0.999,
+        'batch_size': BATCH_SIZE,
+        'num_epochs': 100,
+        'feature_maps': 64,
+        'image_size': 64,
+    }
```
[the remark this answers is not in the document]
Comment (Jschechter18) [reply code/train.py:45]: this is ok if you plan to do another PR for tuning implemenation before training

File: code/train.py [hunk 2]
```diff
@@ -163,24 +179,106 @@ def tune_progan(train_loader, val_loader):
...
+            # Train Generator
+            z = torch.randn(batch_size, LATENT_DIM, 1, 1, device=DEVICE)
+            x_fake = model.generator(z)
+            gen_out = model.discriminator(x_fake)
+
+            gen_loss = criterion(gen_out, real_labels)
+
+            generator_optimizer.zero_grad()
+            gen_loss.backward()
+            generator_optimizer.step()
```
[the remark this answers is not in the document]
Comment (Jschechter18) [reply code/train.py:319]: this is critical. just make sure to address this so we aren't risking using stale gradients downstream

File: code/train.py [hunk 3]
```diff
@@ -163,24 +179,106 @@ def tune_progan(train_loader, val_loader):
...
+    model.to(DEVICE)
+
+    if params.get('start_epoch', 0) == 0:
+        model.apply(weights_init)
+
+    model.train()
+
+    best_gen_loss = float('inf')
+
+    for epoch in range(params.get('start_epoch', 0), params['num_epochs']):
```
[the remark this answers is not in the document]
Comment (Jschechter18) [reply code/train.py:276]: this should be ok as long as tuning implementation is correct. for this PR, consider commenting the result section, then uncommenting once correct tuning function is implemented
